ng-dev/release/notes/changelog.ts

```diff
@@ -0,0 +1,141 @@
+export const changelogPath = 'CHANGELOG.md';
+
+/** Project-relative path for the changelog archive file. */
+export const changelogArchivePath = 'CHANGELOG_ARCHIVE.md';
+
+/** A marker used to split a CHANGELOG.md file into individual entries. */
+export const splitMarker = '<!-- CHANGELOG SPLIT MARKER -->';
+
+/**
+ * A string to use between each changelog entry when joining them together.
+ *
+ * Since all every changelog entry's content is trimmed, when joining back together, two new lines
+ * must be placed around the splitMarker to create a one line buffer around the comment in the
+ * markdown.
+ * i.e.
+ * <changelog entry content>
+ *
+ * <!-- CHANGELOG SPLIT MARKER -->
+ *
+ * <changelog entry content>
+ */
+const joinMarker = `\n\n${splitMarker}\n\n`;
+
+/** A RegExp matcher to extract the version of a changelog entry from the entry content. */
+const versionAnchorMatcher = new RegExp(`<a name="(.*)"></a>`);
+
+/** An individual changelog entry. */
+interface ChangelogEntry {
+  content: string;
+  version: semver.SemVer;
+}
+
+export class Changelog {
+  /** The absolute path to the changelog file. */
+  readonly filePath = join(this.git.baseDir, changelogPath);
+  /** The absolute path to the changelog archive file. */
+  readonly archiveFilePath = join(this.git.baseDir, changelogArchivePath);
+
+  /** The changelog entries in the CHANGELOG.md file. */
+  private entries = this.getEntriesFor(this.filePath);
+  /**
+   * The changelog entries in the CHANGELOG_ARCHIVE.md file.
+   * Delays reading the CHANGELOG_ARCHIVE.md file until it is actually used.
+   */
+  private get archiveEntries() {
+    if (this._archiveEntries === undefined) {
+      return (this._archiveEntries = this.getEntriesFor(this.archiveFilePath));
+    }
+    return this._archiveEntries;
+  }
+  private _archiveEntries: undefined | ChangelogEntry[] = undefined;
+  /** Whether the changelog archive has entries upon intialization. */
+  private alreadyHasChangelogArchive = this.archiveEntries.length !== 0;
+
+  constructor(private git: GitClient) {}
+
+  /** Prepend a changelog entry to the changelog. */
+  prependEntryToChangelog(entry: string) {
+    this.entries.unshift(parseChangelogEntry(entry));
+    this.writeToChangelogFiles();
+  }
+
+  /**
+   * Move all changelog entries from the CHANGELOG.md file for versions prior to the provided
+   * version to the changelog archive.
+   *
+   * Versions should be used to determine which entries are moved to archive as versions are the
+   * most accurate piece of context found within a changelog entry to determine its relationship to
+   * other changelog entries.  This allows for example, moving all changelog entries out of the
+   * main changelog when a version moves out of support.
+   */
+  moveEntriesPriorToVersionToArchive(version: semver.SemVer) {
+    [...this.entries].reverse().forEach((entry: ChangelogEntry) => {
+      if (version.compare(entry.version) === 1) {
```

Comment:
I think we should use: semver.lt or gt as that is more obvious, compared to the current function whic primarily exists for comparators in e.g a sort